### PR TITLE
[UI] add teacher preview mode

### LINKED ISSUE #3374: [UI] "preview as student" option for teachers
**Idea incl level**

It will be a lot of work but it would be great if we could have a "view as student" button in the Teacher's UI where a teacher can see exactly what the interface will look like for students (I do this my opening a second browser but that is kind of cumbersome).

This is quite a common features in such teacher tools.

### PR DESCRIPTION
**Pull request by jtwaleson**
**Description**

Adds a preview mode for teachers, so when they create a class they can view Hedy as if they were a student in this class.

Fixes #3374

This is a draft PR, please let me know if this is roughly the right direction. Once we confirm that I'll add:
- some tests
- make the banner "Preview Mode" look nicer
- add translation keys

**How to test**

Step 1 - Go to For Teacher page. Make sure you have a class with some customizations, like less programs for level 1. Press the "Preview" button.

Screenshot from 2023-11-29 12-09-57

Step 2 - You should go to /hedy and see the loaded class including a banner "Preview mode".

Screenshot from 2023-12-18 13-45-47

The previous version of this design is [here](https://github.com/hedyorg/hedy/assets/1298565/a13ac70d-5524-435a-884a-061164e3160d).

Step 3 - You can complete the programs and puzzles as if you were a student

Step 4 - When you exit preview mode the banner should disappear and you should go back to the /for-teachers page.

**Checklist**
Done? Check if you have it all in place using this list: (mark with x if done)

- [ ] Contains one of the PR categories in the name
- [ ] Describes changes in the format above
- [ ] Links to an existing issue or discussion
- [ ] Has a "How to test" section

### CODE REVIEW

File: templates/layout.html [hunk 1]
```diff
@@ -78,6 +78,12 @@
       {% if menu != False %}
         {% include "incl/menubar.html" %}
       {% endif %}
+      {% if session.preview_class %}
+      <div class="bg-red-100 border-l-4 border-red-500 text-red-700 p-4" role="alert">
+          <p class="font-bold">Preview mode</p>
+          <p>You are previewing a class as a teacher. <a href="/for-teachers/clear-preview-class">Exit preview mode</a></p>
```
Comment (hasan-sh) [inline templates/layout.html:84]: can we add the class name here? Just to know which class you're viewing without having to go to the customization and figure out which one!

File: website/for_teachers.py [hunk 2]
```diff
@@ -208,6 +208,26 @@ def get_class(self, user, class_id):
...
 
+    @route("/class/<class_id>/preview", methods=["GET"])
+    @requires_login
+    def preview_class_as_teacher(self, user, class_id):
+        if not is_teacher(user) and not is_admin(user):
+            return utils.error_page(error=403, ui_message=gettext("retrieve_class_error"))
+        Class = self.db.get_class(class_id)
+        if not Class or (not utils.can_edit_class(user, Class) and not is_admin(user)):
+            return utils.error_page(error=404, ui_message=gettext("no_such_class"))
+        session["preview_class"] = Class["id"]
```
Comment (hasan-sh) [inline website/for_teachers.py:219]: perhaps we should save a dict with `id` and `class_name` class name instead.

File: website/database.py [hunk 3]
```diff
@@ -775,6 +775,11 @@ def get_student_class_customizations(self, user):
         if student_classes:
             class_customizations = self.get_class_customizations(student_classes[0]["id"])
             return class_customizations or {}
+        elif preview_class_as_teacher:
+            for Class in self.get_teacher_classes(user):
+                if preview_class_as_teacher == Class["id"]:
```
Comment (hasan-sh) [inline website/database.py:780]: class_to_preview? because preview_class_as_teacher comes across as a boolean!

File: app.py [hunk 4]
```diff
@@ -1313,7 +1313,8 @@ def index(level, program_id):
 
     customizations = {}
     if current_user()['username']:
-        customizations = DATABASE.get_student_class_customizations(current_user()['username'])
+        customizations = DATABASE.get_student_class_customizations(
+            current_user()['username'], preview_class_as_teacher=session.get("preview_class"))
```
Comment (hasan-sh) [inline app.py:1317]: Should we add it to https://github.com/hedyorg/hedy/blob/main/app.py#L1149 as well?

File: website/database.py (hunk 3 above)
Comment (jtwaleson) [reply website/database.py:780]: Agreed!

File: website/for_teachers.py (hunk 2 above)
Comment (jtwaleson) [reply website/for_teachers.py:219]: Agreed

File: templates/layout.html (hunk 1 above)
Comment (jtwaleson) [reply templates/layout.html:84]: Yes!

File: app.py (hunk 4 above)
Comment (jtwaleson) [reply app.py:1317]: Good point. Does this functionality make sense for the hour of code version? I'm not familiar with how that works.

Comment (hasan-sh) [reply app.py:1317]: Perhaps not, since we don't really require a teacher for that indeed!

**Comment by Felienne** [comment]
Thanks @jtwaleson and sorry for being a bit late with the feedback, I was traveling last 2 weeks. 

I love that we are making this work!! I think the preview looks great, the banner is clear and it is awesome that it is really working where you can see and try examples!

I was wondering though about the placement of the button to reach this feature. The overview page is fine, we can totally leave it there, but... I think the most natural workflow for teachers would be to want to review the class when they are customizing, and not when they are in the overview. 

So can we also add a preview button here?
image

**Comment by Felienne** [comment]
Ow and I think #4870 might cause some conflicts but I can deal with these for you

**Comment by jtwaleson** [comment]
> Thanks @jtwaleson and sorry for being a bit late with the feedback, I was traveling last 2 weeks.
> 
> I love that we are making this work!! I think the preview looks great, the banner is clear and it is awesome that it is really working where you can see and try examples!
> 
> I was wondering though about the placement of the button to reach this feature. The overview page is fine, we can totally leave it there, but... I think the most natural workflow for teachers would be to want to review the class when they are customizing, and not when they are in the overview. Now if I want to preview after customization I have to click back twice, and if I ant to make a change based on the preview, I have to click again twice to reach the customization again.
> 
> So can we also add a preview button here? image

Great, will do! I'll wrap up the new button and the tests this week.

**Comment by jtwaleson** [comment]
@hasan-sh @Felienne ready for review!

**Comment by Felienne** [comment]
Hi @hasan-sh!

Can you review this in terms of the code? I am totally happy with the way it works, if you are ok too, we can ship this!

@jtwaleson the test that you added now break, can you take a look at this?

**Pull request merged**